feat: 풀이 게시글 작성 기능 추가(#27)

handseokjin opened the pull request:
- [x] 🔀 PR 제목의 형식을 잘 작성했나요? e.g. `[feat] PR을 등록한다.` 
- [x] 💯 테스트는 잘 통과했나요?
- [x] 🏗️ 빌드는 성공했나요?
- [x] 🧹 불필요한 코드는 제거했나요?
- [x] 💭 이슈는 등록했나요?
- [x] 🏷️ 라벨은 등록했나요?
- [x] 💻 git rebase를 사용했나요?

## 작업 내용
풀이 게시글 작성 기능 추가
 - 풀이 게시글 테이블 내 데이터 정상 추가
 - 풀이와 프로그래밍 언어를 연결하는 풀이_언어 엔티티 내 데이터 정상 추가

## 스크린샷
- 회원가입
<img width="636" alt="회원가입 완료" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/d695624a-89a0-40c2-bd77-80c79aac9d44"><br/>
- 로그인
<img width="636" alt="로그인 완료" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/7378c446-7471-4be1-b712-fcdc632c44d7"><br/>
- 데이터 우선 등록(문제에 사용가능한 언어 및 문제 데이터)<br/>
<img width="565" alt="문제에 사용가능한 언어 및 문제 데이터 삽입" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/5b10489e-bc2f-4c72-835d-1ce45978066f"><br/>
- 풀이 글 작성 완료
  - 요청 : jwt 토큰이 필요
  - 응답 : 201 응답 코드 및 상세 페이지 정보 요청을 위한 uri를 전달합니다(헤더명 : Location).<br/>
<img width="638" alt="풀이 글 작성 완료" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/21414fd3-e098-4133-9ac0-1675b838ecde"><br/>
<img width="638" alt="풀이 글 정상 작성 완료" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/bc337f04-c18d-4648-b0ee-0fb1154871d1"><br/>
- 풀이 작성 글 및 풀이_언어 데이터 db에 정상 저장 완료<br/>
   - 풀이 작성 글<br/>
<img width="530" alt="풀이 작성 글 mysql" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/5689ba8d-16b2-4511-9bc1-97177801bd3c"><br/>
  - 풀이_언어 <br/>
<img width="477" alt="풀이 언어 정상 동작" src="https://github.com/NaKaLiGoBa/AlgoWithMe-BE/assets/32458465/2dd56b03-deef-4255-845c-c10d74dbb254"><br/>


## 주의사항
 - 회원, 문제, 프로그래밍 언어 테이블에 값을 우선 넣어서 테스트 하였습니다.
   풀이 등록 뿐만 아니라 특정 회원이 특정 문제에서 풀이 가능한 프로그래밍 언어를 풀이_언어에 등록해야 하므로.

Closes #27

Commit:
feat: 풀이 게시글 작성 기능 추가(#27)

handseokjin commented:
풀이 작성 시 풀이 데이터 및 어떤 언어로 작성했는지 저장을 위한 풀이_언어 데이터에도 같이 저장하는 게 맞는지 확인 부탁드립니다(다대다 관계를 해결하기 위한 연결엔티티에도 값을 넣는 게 맞는지)!

acisliver commented on src/main/java/com/nakaligoba/backend/solution/controller/dto/SolutionCreateRequest.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.ArrayList;
+
+@Data
+public class SolutionCreateRequest {
+    private String title;
+    private String content;
+    private ArrayList<String> languages;
+}
```
입력 검증이 되어있지 않습니다.
`@NotBlank`, `@NotEmpty`를 적절히 사용해야 `@Valid` 를 통한 바인딩 에러 처리가 가능합니다!

handseokjin replied:
아, 대은님 감사합니다. 이 부분을 놓쳤습니다.

acisliver commented on src/main/java/com/nakaligoba/backend/solution/domain/Solution.java:
```diff
@@ -38,4 +37,12 @@ public class Solution extends BaseEntity {
+        this.member = member;
+        this.problem = problem;
+    }
 }
```
만약 연관관계 편의 메서드를 만드신다면 `List<SolutionLanguage> solutionLanguages` 배열에 `SolutionLanguage`를 넣는 것이 필요할 것 같습니다. 예를들어
```java
public addSolutionLanguage(ProgrammingLanguage lang) {
    SolutionLanguage sl = new SolutionLanguage(this, lang);
    this.solutionLanguages.add(sl)
}
```

handseokjin replied:
대은님 혹시 말씀하신 게 save 자체를 여러 번 하는 것이 아니라 리스트를 만들어서 한 번에 saveAll 하는 것을 말씀하시는 거죠??

acisliver replied:
솔루션 입장에서 `SolutionLanguage`를 저장한다면 이런 메서드를 두는 게 좋을 것 같아서요! 근데 저희가 new하고 저렇게 넣는것만으로는 영속성 컨텍스트에 `SolutionLanguage`가 저장되지 않긴 하겠네요. 그냥 두시면 될 것 같습니다!

acisliver commented on src/main/java/com/nakaligoba/backend/solution/application/SolutionService.java:
```diff
@@ -0,0 +1,56 @@
+    private final ProgrammingLanguageRepository programmingLanguageRepository;
+    private final SolutionLanguageRepository solutionLanguageRepository;
+
+    public long createSolution(long problemId, SolutionCreateRequest request) {
+        Member member = memberRepository.findByEmail(JwtUtils.getEmailFromSpringSession());
```
이렇게 쓰는 것 보다 Controller가 멤버 이메일을 서비스에 전달해주는 것이 좋을 것 같습니다.
전체 애플리케이션에서 봤을 때는 세션에 들어온 멤버가 솔루션을 생성하는 것이지만,
단일 서비스의 메서드 입장에서는 어떤 멤버가 솔루션을 작성하는지 알 수 없습니다.
이는 테스트하기 어려운 메서드가 되구요!
```suggestion
    public long createSolution(String email, long problemId, SolutionCreateRequest request) {
        Member member = memberRepository.findByEmail(email);
```

handseokjin replied:
아 맞네요. 단일 서비스 + 테스트 관련해서 말씀하신 부분이 맞네요. 이해했습니다. 감사합니다!

acisliver commented:
> 풀이 작성 시 풀이 데이터 및 어떤 언어로 작성했는지 저장을 위한 풀이_언어 데이터에도 같이 저장하는 게 맞는지(다대다 관계를 해결하기 위해 만든 연결 엔티티에도 값을 넣는 게 맞는지) 확인 부탁드립니다!

같이 저장하는 게 맞습니다!

Commit:
chore: 리뷰 부분 수정

 - SolutionCreateRequest 내 검증을 위한 어노테이션 추가
 - SolutionService 내 createSolution() 메서드에 이메일 전달하여 처리
 - 전달된 풀이 글 작성 언어의 개수만큼 save하는 것이 아닌 arrayList를 통해 한 번에 saveAll하도록 수정

acisliver commented on src/main/java/com/nakaligoba/backend/solution/controller/dto/SolutionCreateRequest.java:
```diff
@@ -2,11 +2,18 @@
+    @NotBlank
     private String content;
+
+    @NotNull
```
언어 배열이 비면 안되니까 `NotEmpty`가 더 좋을 것 같습니다!

handseokjin replied:
엇;; 맨 처음에 `NotBlank`로 했다가 안 되길래 ArrayList라서 "" 이런 거는 안 되는 건가 해서 `NotNull`만 했는데 `NotEmpty는 또 되는군요.. 감사합니다!

Commit:
style: SolutionCreateRequest 내 languages 변수 어노테이션 NotEmpty로 수정

handseokjin commented:
감사합니다. 리뷰해 주시는 덕분에 항상 많이 배우는 것 같습니다! 다음에는 최대한 번거롭지 않게 해드릴 수 있기를..!
